File: go+/0209M-长度最小的子数组.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-12 09:10:58
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func minSubArrayLen(target int, nums []int) int {
	l := 0
	sum, ans := 0, 100005
	for r := range nums {
		sum += nums[r]
		for sum >= target {
			if r-l+1 < ans {
				ans = r - l + 1
			}
			sum -= nums[l]
			l++
		}
	}
	if ans == 100005 {
		return 0
	}
	return ans
}

// 时间 O(n)
// 空间 O(1)

File: redo/0209M-长度最小的子数组.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-02-04 20:18:54
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int ans = INT32_MAX;
        int sum = 0;
        for (int l = 0, r = 0; r < nums.size(); r++) {
            sum += nums[r];
            if (sum < target) continue;
            while (sum - nums[l] >= target) {
                sum -= nums[l++];
            }
            ans = min(ans, r - l + 1);
        }
        if (ans == INT32_MAX) return 0;
        return ans;
    }
};